### ISSUE TITLE
[Feature Request] integrating intel extension for pytorch

### ISSUE THREAD

**Issue by george-adams1** [DLR-RM/stable-baselines3#1564]
### 🚀 Feature

Request to integrate the Intel extension for PyTorch into sb3. The Intel extension for PyTorch optimizes the PyTorch library to better utilize the computational capabilities of Intel processors. By integrating this extension, SB3 users that utilize Intel processors could potentially experience significant performance improvements.

### Motivation

Maximizing the performance of Intel processors. This can lead to faster training times and more efficient resource utilization, improving the overall experience for users of SB3 who use Intel processors.

### Pitch

The integration of the Intel extension for PyTorch into SB3 would involve modifying the library to utilize the extension when it detects that it is running on an Intel processor. 

### Alternatives

An alternative to this would be to provide guidelines on how users can manually integrate the Intel extension into their SB3 setups. 

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo

**Comment by araffin** [comment]
Hello,
could you give some pointer/example on how to integrate that extension?

will you be willing to contribute such extension?

**Comment by duburcqa** [comment]
I believe it was a reference to this project: https://github.com/intel/intel-extension-for-pytorch

The latest version is compatible with `torch.compile`. This means that it is now much easier to leverage their optimizations without a complete rework of the libraries relying on pytorch. Here is the example script https://github.com/intel/intel-extension-for-pytorch/blob/7d205d613130c1b344a2f3722d8894ab1bf2264d/examples/cpu/inference/python/bert_torchdynamo_mode_inference_fp32.py#L16

**Comment by araffin** [comment]
Thanks for the pointers @duburcqa =)!

I see. In that case, it is already supported as SB3 exposes its PyTorch policy via the `.policy` attribute (this is a `nn.Module`, see doc).

Related: https://github.com/DLR-RM/stable-baselines3/pull/1391 and https://github.com/DLR-RM/stable-baselines3/pull/1439 

You should also know that the bottleneck at train time is the gradient update, so it won't help you much to optimize inference time (unless you are at test time), we do have an experimental Jax version (SBX) if you need significant speed up (see README).

**Comment by araffin** [comment]
I would actually welcome a PR that shows how to use `th.compile` in our doc =)

**Comment by araffin** [comment]
Ad a follow up, it would be nice to investigate where to have torch compile for some speed boost (see https://github.com/pytorch-labs/LeanRL, cc @vmoens).
Given SBX experience (https://github.com/araffin/sbx), biggest speed boost should come from compiling the backward pass and combining multiple gradient steps to avoid for loops and minimizing transfer between CPU/GPU (to see the speed boost, maybe starting with CrossQ from SB3 contrib).

**Comment by vmoens** [comment]
Hey
Happy to help with this.
Compiling several steps together isn't a bad idea. 
The only thing that usually works best is this:
- put all the updates in a single `update` function, including zero_grad, forward, backward, clip grad norm, step and target params updates
- if possible group optimizers together (this can be done using param_groups in they all have the same number of steps)
- have a look at the graph breaks using `TORCH_LOGS="+graph_breaks,recompiles" python code.py`
- avoid cpu-gpu transfers in the compiled region, and use torch whenever possible (not numpy).
- prefer torch.where to masking, avoid control flow if can be avoided (again sometimes torch.where can help)
- use mode="reduce-overhead" or tensordict's CudaGraphModule like we do in leanrl 

That's the 101 but I'd be happy to review or help implement a compiled version of any script.

**Comment by sdace9719** [comment]
I would like to work on this, if possible.

**Comment by sdace9719** [comment]
I am not sure which is the right section in the documentation to put this. Should I create a new section for this?

**Comment by araffin** [comment]
> I am not sure which is the right section in the documentation to put this. Should I create a new section for this?

At first, you can add it to the "Examples".

**Comment by sdace9719** [comment]
I added a PR with doc changes.

**Comment by sdace9719** [comment]
I see this as closed so do we have any plans to add torch.compile in SB3 for a speed boost

**Comment by araffin** [comment]
> I see this as closed so do we have any plans to add torch.compile in SB3 for a speed boost

no direct plans currently no. 
For speed, I would usually recommend to use SBX (SB3 + Jax) also it has less features.
